pccommon/pccommon/middleware.py

```diff
@@ -61,3 +65,31 @@ async def dispatch(
         await self.set_body(request)
         response = await trace_request(self.service_name, request, call_next)
         return response
+
+
+async def timeout_middleware(
+    request: Request,
+    call_next: Callable[[Request], Awaitable[Response]],
+    timeout: int,
+) -> Response:
+    try:
+        start_time = time.time()
+        return await asyncio.wait_for(call_next(request), timeout=timeout)
+
+    except asyncio.TimeoutError:
+        process_time = time.time() - start_time
+        log_dimensions = get_custom_dimensions({"request_time": process_time}, request)
+
+        logger.exception(
+            "Request timeout",
+            extra=log_dimensions,
+        )
+
+        ref_id = log_dimensions["custom_dimensions"].get("ref_id")
+        debug_msg = f"Debug information for support: {ref_id}" if ref_id else ""
+
+        return PlainTextResponse(
+            f"The request exceeded the maximum allowed time, please try again."
+            f"\n\n{debug_msg}",
+            status_code=HTTP_504_GATEWAY_TIMEOUT,
```

Comment:
For consideration: 504 seems mostly appropriate, however this is not a "gateway" timeout necessarily. It may be that we also want to distinguish this error from the 504s that Frontdoor also seems to return. Perhaps this should just be a 500?

Reply:
Would we be able to differentiate based on response content in the logs? If so 504 seems right...500 might be more technically correct, but I think it's better to differentiate between exceptions and timeouts.

Reply:
Yes, mostly. We currently have no 504 or 503 return types in the "api requests" log since those errors are currently circumventing our batched Azure logging handler. In the Frontdoor logs it may be more difficult to differentiate, but that should be less of an issue since (hopefully) we'll now get the api request and trace logs (both tagged with the ref_id) in our application monitoring. I'll leave it at 504 now and make sure the logs make sense when deployed to staging.